### PR TITLE
feat(protocol-designer): mixpanel events pt2 redesign

### PR DESCRIPTION
**Pull request by jerader**
closes AUTH-1017

# Overview

Refactor the pipetting args and add an event in for formWarnings and formErrors

## Test Plan and Hands on Testing

Review the code, i checked that it works in mixpanel correctly!

## Changelog

- add in event for formErrors and formWarnings
- refactor pipetting args events

## Risk assessment

low

### CODE REVIEW

File: protocol-designer/src/analytics/middleware.ts [hunk 1]
```diff
@@ -39,6 +50,19 @@ import type { SetFeatureFlagAction } from '../feature-flags/actions'
 import type { CreatePipettesAction } from '../step-forms/actions'
 import type { AnalyticsEventAction } from './actions'
 import type { AnalyticsEvent } from './mixpanel'
+
+const PIPETTING_ARGS_FILTER_LIST = [
```
Comment (koji) [inline protocol-designer/src/analytics/middleware.ts:55]: what does the strikethrough mean in the spreadsheet?
https://docs.google.com/spreadsheets/d/1AiFdoyO0M4r3eDJ6s9Qpvqk0rCt2P-HWdQjqLXmU5Vs/edit?gid=1585271769#gid=1585271769

Comment (jerader) [reply protocol-designer/src/analytics/middleware.ts:55]: it means we don't want to record the field values for some reason (or we want to rename it) - Felix decided on the list

File: protocol-designer/src/analytics/middleware.ts [hunk 2]
```diff
@@ -123,6 +147,94 @@ export const reduxActionToAnalyticsEvent = (
           name: 'heaterShakerStep',
           properties: {},
         }
+      } else if (
```
Comment (koji) [inline protocol-designer/src/analytics/middleware.ts]: nit maybe switch statement is a little bit easy to read?

File: protocol-designer/src/analytics/middleware.ts [hunk 3]
```diff
@@ -123,6 +147,94 @@ export const reduxActionToAnalyticsEvent = (
...
+          name: `mixStep`,
+          properties: {
+            ...stepArgModified,
+            aspirateFlowRate: stepArgModified.aspirateFlowRateUlSec,
+            dispenseFlowRate: stepArgModified.dispenseFlowRateUlSec,
+            blowoutFlowRate: stepArgModified.blowoutFlowRateUlSec,
+            aspirateOffsetFromBottomMm:
+              stepArgModified.aspirateOffsetFromBottomMm ===
+              DEFAULT_MM_FROM_BOTTOM_ASPIRATE
+                ? 'default'
```
Comment (koji) [inline protocol-designer/src/analytics/middleware.ts]: nit
```ts
const DEFAULT_VALUE = 'default'

? DEFAULT_VALUE : stepArgModified.aspirateOffsetFromBottomMm
```

File: protocol-designer/src/pages/Designer/ProtocolSteps/StepForm/StepFormToolbox.tsx [hunk 4]
```diff
@@ -157,6 +157,28 @@ export function StepFormToolbox(props: StepFormToolboxProps): JSX.Element {
...
 
+  const visibleFormWarningsTypes = visibleFormWarnings.map(
+    warning => warning.type
+  )
+  const visibleFormErrorsTypes = visibleFormErrors.map(error => error.title)
+
+  useEffect(() => {
+    if (visibleFormWarningsTypes.length > 0) {
+      const formWarningsEvent: AnalyticsEvent = {
+        name: 'formWarnings',
```
Comment (koji) [inline protocol-designer/src/pages/Designer/ProtocolSteps/StepForm/StepFormToolbox.tsx]: nit
we define event name as a const in app. probably using the same way would be good.

app/src/redux/analytics/constants.ts

File: protocol-designer/src/pages/Designer/ProtocolSteps/StepForm/StepFormToolbox.tsx [hunk 5]
```diff
@@ -157,6 +161,28 @@ export function StepFormToolbox(props: StepFormToolboxProps): JSX.Element {
...
+      dispatch(analyticsEvent(formWarningsEvent))
+    }
+    if (visibleFormErrorsTypes.length > 0) {
+      const formErrorsEvent: AnalyticsEvent = {
+        name: FORM_ERRORS_EVENT,
+        properties: { visibleFormErrorsTypes },
+      }
+      dispatch(analyticsEvent(formErrorsEvent))
+    }
+  }, [visibleFormWarningsTypes, visibleFormErrorsTypes])
```
Comment (koji) [inline protocol-designer/src/pages/Designer/ProtocolSteps/StepForm/StepFormToolbox.tsx]: nit 

```ts
  useEffect(() => {
    const dispatchAnalyticsEvent = (
      eventName: string,
      eventProperties: FormWarningType[] | string[]
    ): void => {
      if (eventProperties.length > 0) {
        const event: AnalyticsEvent = {
          name: eventName,
          properties: { eventProperties },
        }
        dispatch(analyticsEvent(event))
      }
    }

    dispatchAnalyticsEvent(FORM_WARNINGS_EVENT, visibleFormWarningsTypes)
    dispatchAnalyticsEvent(FORM_ERRORS_EVENT, visibleFormErrorsTypes)
  }, [visibleFormWarningsTypes, visibleFormErrorsTypes])

```

**Review by koji — APPROVED** [review]
lgtm
just left one comment

File: protocol-designer/src/pages/Designer/ProtocolSteps/StepForm/StepFormToolbox.tsx [hunk 6]
```diff
@@ -157,6 +162,29 @@ export function StepFormToolbox(props: StepFormToolboxProps): JSX.Element {
...
+          name: eventName,
+          properties: { eventProperties },
+        }
+        dispatch(analyticsEvent(event))
+      }
+    }
+
+    dispatchAnalyticsEvent(FORM_WARNINGS_EVENT, visibleFormWarningsTypes)
+    dispatchAnalyticsEvent(FORM_ERRORS_EVENT, visibleFormErrorsTypes)
+  }, [visibleFormWarningsTypes, visibleFormErrorsTypes])
```
Comment (koji) [inline protocol-designer/src/pages/Designer/ProtocolSteps/StepForm/StepFormToolbox.tsx:186]:
```suggestion
  }, [visibleFormWarningsTypes, visibleFormErrorsTypes, dispatch])
```

Comment (jerader) [reply protocol-designer/src/pages/Designer/ProtocolSteps/StepForm/StepFormToolbox.tsx:186]: not sure adding dispatch to the array is needed

if you look at the first comment in here: https://stackoverflow.com/questions/77293395/why-should-i-pass-in-dispatch-as-a-dependency-of-useeffect-in-react

**Pull request merged**